pkg/storaged/nfs-mounts.py

```diff
@@ -0,0 +1,344 @@
+#! /usr/bin/python
+
+# nfs-mounts   --  Monitor and manage NFS mounts
+#
+# This is similar to how UDisks2 monitors and manages block device
+# mounts, but for NFS.  This might be moved into UDisks2, or not.
+
+import select
+import re
+import sys
+import json
+
+import os
+import ctypes
+import struct
+import subprocess
+import traceback
+
+# We monitor all NFS remotes listed in /etc/fstab and in /etc/mtab.
+# If a entry from mtab is also found in fstab, we report only the
+# fstab entry and mark it is "mounted".
+
+IN_CLOSE_WRITE = 0x00000008
+IN_IGNORED     = 0x00008000
+
+class Inotify:
+    def __init__(self):
+        self._libc = ctypes.CDLL(None, use_errno=True)
+        self._get_errno_func = ctypes.get_errno
+
+        self._libc.inotify_init.argtypes = []
+        self._libc.inotify_init.restype = ctypes.c_int
+        self._libc.inotify_add_watch.argtypes = [ctypes.c_int, ctypes.c_char_p,
+                                                 ctypes.c_uint32]
+        self._libc.inotify_add_watch.restype = ctypes.c_int
+        self._libc.inotify_rm_watch.argtypes = [ctypes.c_int, ctypes.c_int]
+        self._libc.inotify_rm_watch.restype = ctypes.c_int
+
+        self.fd = self._libc.inotify_init()
+
+    def add_watch(self, path, mask):
+        path = ctypes.create_string_buffer(path.encode(sys.getfilesystemencoding()))
+        wd = self._libc.inotify_add_watch(self.fd, path, mask)
+        if wd < 0:
+            sys.stderr.write("can't add watch for %s: %s\n" % (path, os.strerror(self._get_errno_func())))
+        return wd
+
+    def rem_watch(self, wd):
+        if self._libc.inotify_rm_watch(self.fd, wd) < 0:
+            sys.stderr.write("can't remove watch: %s\n" % (os.strerror(self._get_errno_func())))
+
+    def process(self, callback):
+        buf = os.read(self.fd, 4096)
+        pos = 0
+        while pos < len(buf):
+            (wd, mask, cookie, name_len) = struct.unpack('iIII', buf[pos:pos+16])
+            pos += 16
+            (name,) = struct.unpack('%ds' % name_len, buf[pos:pos + name_len])
+            pos += name_len
+            callback(wd, mask, name.decode().rstrip('\0'))
```

Comment:
IMHO this code is too obscure and finnicky to carry it around multiple times in the source. Could the inotify bits move to pkg/lib/ and get included (statically during build) in both applications and storage? Of course a "proper" implementation in the bridge would be better in the long run (extending our `file.watch()` API), but we can't rely on that just yet on e. g. RHEL.

Reply:
I think we can require changes to the bridge for a new feature.  The shell and dashboard need to be compatible with all versions of bridges, but a page can always bring exactly the bridge it needs, no?

However, improving the bridge doesn't help a Python script.  We would have to reimplement it all in JavaScript and run it in the browser.  This is possible of course, but has no advantages except being able to use the bridge APIs.  (Writing synchronous code in Python is sooo much cleaner than writing asynch code in JavaScript, and we also avoid a ton of round trips.)

Which means, yes, I should move the inotify bits out into pkg/lib.

Reply:
Why does it need to happen in Java? The bridge already provides the inotify ↔ channel implementation, so that could be extended to be able to watch for directories too (what's done in Python). Then JS should only use the `file.watch()` API.

However:
> a page can always bring exactly the bridge it needs, no?

Not for RHEL 7.4 - there, the bridge is shipped in base and cannot (easily) change, but in extras we ship cockpit-{storaged,packagekit} which get rebased frequently. For that case we need to provide a compatibility shim if we want to support it there, at least until the rebase in 7.5. Or just don't support this feature until 7.5, which sounds perfectly acceptable, too.

Reply:
> Not for RHEL 7.4 - there, the bridge is shipped in base and cannot (easily) change, but in extras we ship cockpit-{storaged,packagekit} which get rebased frequently.

I see.  This is another argument against including remote mounts in our fsys sampling.

Reply:
> Why does it need to happen in Java? The bridge already provides the inotify ↔ channel implementation, so that could be extended to be able to watch for directories too (what's done in Python). Then JS should only use the file.watch() API.

Hmm, I don't follow.  In this PR, we don't need to watch directories.  Are you thinking of `pkg/apps`?  Here, we need to watch files in `/proc`, which don't work with `inotify` but they do work with `select`.

We can extend the fswatch1 payload type to also do `proc` / `sysfs` style change notifications.  (And we should in any case.)

But don't we then need to write all the logic that is now in `nfs-mounts.py` in JavaScript and run it in the browser?  Or do you want to do the file watching in the browser, and the fstab parsing etc in Python?

(I actually started out with implementing this in JavaScript, when I thought I don't need to watch `/proc/self/mounts`, but I quickly got too irritated.)

Reply:
The inotify interface looked pretty much the same, so that should be split out. And you just did, thanks! AFAICS you still use `select()` on an inotify fd (which is the standard way how to do it), so it doesn't work fundamentally differently than in apps (there it just uses a blocking `read()` on the inotify socket, but there is no reason to not use `select()` there also). Not having to watch directories is a difference, I missed that.

So I still wonder how inotify-watching a file in /proc fundamentally differs from watching a file in /usr? And could that not be done much more cleanly in the bridge (in the future)? Then nfs-mounts.py would just do the actions, not the watching? These two parts should be reasonably independent from each other, aren't they?

The "action" parts don't seem to be much more difficult in JS than in Python at first sight, but you worked on this much more closely. I don't mind a Python shim if that makes things easier.

So in summary, I'm happy with the watching in Python as an initial solution (particularly now that the duplicate code got refactored), I'm just thinking aloud how this should look in the future.

Reply:
> So I still wonder how inotify-watching a file in /proc fundamentally differs from watching a file in /usr?

In a nut shell, files in `/proc` and `/sys` don't ever generate inotify events.  See for example http://www.spinics.net/lists/linux-fsdevel/msg73924.html

Reply:
Indeed I was missing `select.select([ fstab_watcher.inotify.fd ], [ ], [ mtab_file ])`, which watches those separately. So yes, using the python wrapper for this bit seems fine for now. Thanks for bearing with me!